conference/cart.py

```diff
@@ -0,0 +1,292 @@
+
+    if request.method == 'POST':
+        stripe_payment = StripePayment.objects.create(
+            order=order,
+            user=request.user,
+            uuid=str(uuid.uuid4()),
+            amount=order.total(),
+            token=request.POST.get('stripeToken'),
+            token_type=request.POST.get('stripeTokenType'),
+            email=request.POST.get('stripeEmail'),
+        )
+        try:
+            with transaction.atomic():
+                charge_for_payment(stripe_payment)
+                order.payment_date = date.today()
+                order.save()
+                create_invoices_for_order(order)
+                # TODO send_confirmation_email()
+
+                return redirect(
+                    'cart:step5_congrats_order_complete', order_uuid=order.uuid
+                )
+        except PaymentError:
+            # Redirect to the same page, show information about failed
+            # payment(s) and reshow the same Pay with Card button
+            return redirect('.')
+
+    return TemplateResponse(
+        request,
+        "ep19/bs/cart/step_4_payment.html",
+        {
+            'order': order,
+            'payments': payments,
+            'total_for_stripe': total_for_stripe,
+        },
+    )
+
+
+@login_required
+def cart_step5_congrats_order_complete(request, order_uuid):
+    order = get_object_or_404(Order, uuid=order_uuid)
+    return TemplateResponse(
+        request,
+        "ep19/bs/cart/step_5_congrats_order_complete.html",
+        {'order': order},
+    )
+
+
+def extract_order_parameters_from_request(post_data):
+    discount_code = None
+    fares_info = {}
+
+    for k, v in post_data.items():
+        if k == 'discount_code':
+            discount_code = v
+
+        elif is_available_fare(k):
+            try:
+                fares_info[k] = int(v)
+            except ValueError:
+                pass
+
+    return discount_code, fares_info
+
+
+def is_available_fare(fare_code):
+    return True
+
+
+class TicketType:
+    personal = 'personal'
+    company = 'company'
+    student = 'student'
+
+    ALL = [personal, company, student]
+
+
+class CartActions:
+    apply_discount_code = 'apply_discount_code'
+    buy_tickets = 'buy_tickets'
+
+
+# TODO: move to fares
+def get_available_fares_for_type(type_of_tickets):
+    assert type_of_tickets in TicketType.ALL
+
+    fares = get_available_fares(datetime.now().date())
+
+    if type_of_tickets == TicketType.personal:
+        regex_group = FARE_CODE_GROUPS.PERSONAL
+    if type_of_tickets == TicketType.company:
+        regex_group = FARE_CODE_GROUPS.COMPANY
+    if type_of_tickets == TicketType.student:
+        regex_group = FARE_CODE_GROUPS.STUDENT
+
+    fares = fares.filter(code__regex=FARE_CODE_REGEXES["groups"][regex_group])
+
+    return fares
+
+
+class PersonalBillingForm(forms.ModelForm):
+    class Meta:
+        model = Order
+        fields = ['card_name', 'country', 'address', 'billing_notes']
```

Comment:
The `card name` field header can be misleading, if I understand correctly, we're asking for `Cardholder's name` here, right?